### PR TITLE
[vpa] update VPA from 0.14 to 1.2.1

### PR DESCRIPTION
**Pull request by borg-z**
## Description

Vertical Pod Autoscaler updated to version 1.2.1
This brought several bug fixes, features, and security patches.

## What is the expected result?

Kubernetes version 1.25 is now supported.

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: vertical-pod-autoscaler
type: feature
summary: Vertical Pod Autoscaler updated to version 1.2.1
impact_level: low
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/static'...

**Comment by borg-z** [comment]
/e2e/abort static;Static;containerd;1.27 11325312685 11325312685-1-con-1-27 /sys/deckhouse-oss/install:pr10097 astra@192.168.199.39

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by borg-z** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.27 11388058303 11388058303-1-con-1-27 /sys/deckhouse-oss/install:pr10097 redos@89.169.138.149 -

**Comment by borg-z** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.27 11363652879 11363652879-1-con-1-27 /sys/deckhouse-oss/install:pr10097 redos@89.169.146.77

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by borg-z** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.27 11455313123 11455313123-1-con-1-27 /sys/deckhouse-oss/install:pr10097 redos@89.169.138.138

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/yandex-cloud'...

File: modules/302-vertical-pod-autoscaler/images/vertical-pod-autoscaler/werf.inc.yaml [hunk 1]
```diff
@@ -9,22 +8,38 @@ git:
...
-    - cd ../../../vertical-pod-autoscaler/pkg/updater
-    - GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -ldflags="-s -w" -o /updater
-    - chown 64535:64535 /updater
-    - chmod 0700 /updater
+    - git clone --depth 1 -b vertical-pod-autoscaler-1.2.1 {{ $.SOURCE_REPO }}/kubernetes/autoscaler.git /src/autoscaler
+    - cd /src/autoscaler
+    - git apply /patches/*.patch --verbose
+    - mv vertical-pod-autoscaler /src/vertical-pod-autoscaler
+    - rm -rf /src/autoscaler
+    - rm -rf /src/vertical-pod-autoscaler/e2e/
```
Comment (mvasl) [inline modules/302-vertical-pod-autoscaler/images/vertical-pod-autoscaler/werf.inc.yaml:16]: What about `/patches`?

File: modules/302-vertical-pod-autoscaler/images/vertical-pod-autoscaler/werf.inc.yaml [hunk 2]
```diff
@@ -9,22 +8,38 @@ git:
...
+    to: /src
+    before: install
+mount:
+  - fromPath: ~/go-pkg-cache
+    to: /go/pkg
+shell:
+  install:
+    - export GO_VERSION=${GOLANG_VERSION} GOPROXY={{ $.GOPROXY }} CGO_ENABLED=0 GOOS=linux GOARCH=amd64
+    - cd /src/vertical-pod-autoscaler/
+    - go mod tidy
```
Comment (mvasl) [inline modules/302-vertical-pod-autoscaler/images/vertical-pod-autoscaler/werf.inc.yaml:32]: `go mod tidy` might change dependecies without notice, `go mod download` is the way

Comment (borg-z) [reply modules/302-vertical-pod-autoscaler/images/vertical-pod-autoscaler/werf.inc.yaml]: fixed

File: modules/302-vertical-pod-autoscaler/images/vertical-pod-autoscaler/werf.inc.yaml (hunk 1 above)
Comment (borg-z) [reply modules/302-vertical-pod-autoscaler/images/vertical-pod-autoscaler/werf.inc.yaml]: Do you think we should do something like rm -rf /patches ?

Comment (mvasl) [reply modules/302-vertical-pod-autoscaler/images/vertical-pod-autoscaler/werf.inc.yaml]: Yep

**Review by mvasl — APPROVED** [review]

**Comment by borg-z** [comment]
HOW-TO test:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test-vpa-deployment
  namespace: default
spec:
  replicas: 1
  selector:
    matchLabels:
      app: test-vpa
  template:
    metadata:
      labels:
        app: test-vpa
    spec:
      containers:
      - name: stress-container
        image: polinux/stress
        resources:
          requests:
            cpu: "50m"
            memory: "100Mi"
          limits:
            cpu: "500m"
            memory: "1Gi"
        command: ["stress"]
        args: ["--cpu", "2", "--vm", "2", "--vm-bytes", "300M", "--timeout", "600s"]
---
apiVersion: autoscaling.k8s.io/v1
kind: VerticalPodAutoscaler
metadata:
  name: test-vpa
  namespace: default
spec:
  targetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: test-vpa-deployment
  updatePolicy:
    updateMode: Auto
  resourcePolicy:
    containerPolicies:
    - containerName: stress-container
      minAllowed:
        cpu: "50m"
        memory: "100Mi"
      maxAllowed:
        cpu: "1000m"
        memory: "2Gi"
```

After some time the status of VerticalPodAutoscaler updates and the pod starts with new resources limits

`kubectl get vpa test-vpa -n default -o yaml`

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @borg-z. I've started the workflow for label 'e2e/run/openstack'...

**Comment by borg-z** [comment]
1. Creating deployment and VPA

```
kubectl apply -f - <<EOF
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test-vpa-deployment
  namespace: default
spec:
  replicas: 2
  strategy:
    type: RollingUpdate
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 0
  selector:
    matchLabels:
      app: test-vpa
  template:
    metadata:
      labels:
        app: test-vpa
    spec:
      tolerations:
        - operator: Exists
      containers:
      - name: stress-container
        image: polinux/stress
        resources:
          requests:
            cpu: "50m"
            memory: "100Mi"
          limits:
            cpu: "500m"
            memory: "1Gi"
        command: ["stress"]
        args: ["--cpu", "2", "--vm", "2", "--vm-bytes", "300M", "--timeout", "600s"]
---
apiVersion: autoscaling.k8s.io/v1
kind: VerticalPodAutoscaler
metadata:
  name: test-vpa
  namespace: default
spec:
  targetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: test-vpa-deployment
  updatePolicy:
    updateMode: Auto
  resourcePolicy:
    containerPolicies:
    - containerName: stress-container
      minAllowed:
        cpu: "50m"
        memory: "100Mi"
      maxAllowed:
        cpu: "1000m"
        memory: "2Gi"
EOF
```

2. Initial state

```
kubectl get pods
NAME                                   READY   STATUS    RESTARTS   AGE
test-vpa-deployment-6c46c7b7cd-nngbf   1/1     Running   0          16s
test-vpa-deployment-6c46c7b7cd-wdzxp   1/1     Running   0          16s

kubectl get pods test-vpa-deployment-6c46c7b7cd-nngbf -o json | jq '.spec.containers[0].resources'
{
  "limits": {
    "cpu": "500m",
    "memory": "1Gi"
  },
  "requests": {
    "cpu": "50m",
    "memory": "100Mi"
  }
}
```

3. VPA recommendation update

```
kubectl describe VerticalPodAutoscaler test-vpa | grep Recommendation: -A100
  Recommendation:
    Container Recommendations:
      Container Name:  stress-container
      Lower Bound:
        Cpu:     522m
        Memory:  488790571
      Target:
        Cpu:     587m
        Memory:  716711186
      Uncapped Target:
        Cpu:     587m
        Memory:  716711186
      Upper Bound:
        Cpu:     1
        Memory:  2Gi
Events:          <none>
```

4. Pod has been recreated

```
kubectl get pods test-vpa-deployment-6c46c7b7cd-2pqdz -o json | jq '.spec.containers[0].resources'
{
  "limits": {
    "cpu": "5870m",
    "memory": "7339122544"
  },
  "requests": {
    "cpu": "587m",
    "memory": "716711186"
  }
}

kubectl get events --sort-by='.lastTimestamp' | grep  wdzxp | grep VPA
6m9s        Normal    EvictedByVPA              pod/test-vpa-deployment-6c46c7b7cd-wdzxp                        Pod was evicted by VPA Updater to apply resource recommendation.
```

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Lada7878. I've started the workflow for label 'deploy/web/stage'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Lada7878. I've started the workflow for label 'deploy/web/test'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Zhbert. I've started the workflow for label 'deploy/web/test'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...

**Review by sprait — APPROVED** [review]

**Review by name212 — APPROVED** [review]

**Pull request merged**